src/mongo/db/pipeline/expression.cpp

```diff
@@ -2716,31 +2731,100 @@ Value ExpressionIndexOfArray::evaluate(const Document& root) const {
+        if (getExpressionContext()->getValueComparator().evaluate(array[i] == args.targetOfSearch)) {
+            return Value(static_cast<int>(i));
+        }
+    }
 
-    Value searchItem = vpOperand[1]->evaluate(root);
+    return Value(-1);
+}
 
-    size_t startIndex = 0;
-    if (vpOperand.size() > 2) {
-        Value startIndexArg = vpOperand[2]->evaluate(root);
+ExpressionIndexOfArray::Arguments ExpressionIndexOfArray::evaluateAndValidateArguments(const Document& root,
+                                                const ExpressionVector& operands,
+                                                size_t arrayLength) const {
+    
+    int startIndex = 0;
+    if (operands.size() > 2) {
+        Value startIndexArg = operands[2]->evaluate(root);
         uassertIfNotIntegralAndNonNegative(startIndexArg, getOpName(), "starting index");
-        startIndex = static_cast<size_t>(startIndexArg.coerceToInt());
+        startIndex = static_cast<int>(startIndexArg.coerceToInt());
     }
+    
 
-    size_t endIndex = array.size();
-    if (vpOperand.size() > 3) {
-        Value endIndexArg = vpOperand[3]->evaluate(root);
+    int endIndex = arrayLength;
+    if (operands.size() > 3) {
+        Value endIndexArg = operands[3]->evaluate(root);
         uassertIfNotIntegralAndNonNegative(endIndexArg, getOpName(), "ending index");
         // Don't let 'endIndex' exceed the length of the array.
-        endIndex = std::min(array.size(), static_cast<size_t>(endIndexArg.coerceToInt()));
+        endIndex = std::min(arrayLength, static_cast<size_t>(endIndexArg.coerceToInt()));
+    }
+    
+    return ExpressionIndexOfArray::Arguments(vpOperand[1]->evaluate(root), startIndex, endIndex);
+}
+/**
+ * This class handles the case where IndexOfArray is given an ExpressionConstant
+ * instead of using a vector and searching through it we can use a unordered_map
+ * for O(1) lookup time.
+ */
+class ExpressionIndexOfArray::Optimized : public ExpressionIndexOfArray {
+public:
+    Optimized(const boost::intrusive_ptr<ExpressionContext>& expCtx,
+              const ValueUnorderedMap<int>& indexMap,
+              const ExpressionVector& operands)
+        : ExpressionIndexOfArray(expCtx), _indexMap(std::move(indexMap)) {
+        vpOperand = operands;
     }
 
-    for (size_t i = startIndex; i < endIndex; i++) {
-        if (getExpressionContext()->getValueComparator().evaluate(array[i] == searchItem)) {
-            return Value(static_cast<int>(i));
+    virtual Value evaluate(const Document& root) const {
+        auto args = evaluateAndValidateArguments(root, vpOperand, _indexMap.size());
+        auto index = _indexMap.find(args.targetOfSearch);
+        if (index != _indexMap.end() && index->second >= args.startIndex &&
+            index->second < args.endIndex) {
+            return Value(index->second);
+        } else {
+            // If the item we are searching is not found or if not found in given range return -1.
```

Comment:
Hmm I actually think this won't work. Consider the following:
```
{$indexOfArray: [[0, 1, "a", 0, 1, 1, "a"], "a", 3, 8]}
```
I think that the map here will be populated with `0-> 0, 1-> 1, "a" -> 2`, but then it has lost the information about where the second or third occurrences of those values are. I think we need to map a value to a `vector<int>` of all the positions so that we can rule out those which are out of range.